File: core/frontend/src/api/modules/emailMedia.ts
import { AxiosProgressEvent } from 'axios'
import { instance } from '@/api'

export interface MediaItem {
	name: string
	url: string
	size: number
	update_time?: number
}

/**
 * @description Upload one image to the shared email-media library.
 * Resolves to the stored { name, url, size }.
 */
export const uploadEmailMedia = (file: File, onProgress?: (percent: number) => void) => {
	const form = new FormData()
	form.append('file', file)
	return instance.post('/email-media/upload', form, {
		headers: { 'Content-Type': 'multipart/form-data' },
		onUploadProgress: (e: AxiosProgressEvent) => {
			if (e.total) onProgress?.(Math.round((e.loaded / e.total) * 100))
		},
	})
}

/**
 * @description List uploaded email images (newest first).
 */
export const listEmailMedia = () => {
	return instance.get('/email-media/list')
}

/**
 * @description Delete an uploaded image by its stored name.
 */
export const deleteEmailMedia = (name: string) => {
	return instance.post(
		'/email-media/delete',
		{ name },
		{ fetchOptions: { successMessage: true } }
	)
}

File: core/api/email_media/v1/email_media.go
package v1

import (
	"billionmail-core/utility/types/api_v1"

	"github.com/gogf/gf/v2/frame/g"
	"github.com/gogf/gf/v2/net/ghttp"
)

// MediaItem describes one uploaded media file.
type MediaItem struct {
	Name       string `json:"name" dc:"Stored file name"`
	Url        string `json:"url" dc:"Public URL to use in emails"`
	Size       int64  `json:"size" dc:"File size in bytes"`
	UpdateTime int64  `json:"update_time" dc:"Last modified (unix seconds)"`
}

// UploadReq uploads a single image to the shared email-media library.
type UploadReq struct {
	g.Meta        `path:"/email-media/upload" method:"post" tags:"EmailMedia" summary:"Upload an email image" mime:"multipart/form-data" in:"body"`
	Authorization string             `json:"authorization" dc:"Authorization" in:"header"`
	File          *ghttp.UploadFile  `p:"file" type:"file" dc:"Image file (png/jpg/gif/webp/bmp)"`
}

// UploadRes returns the public URL of the uploaded image.
type UploadRes struct {
	api_v1.StandardRes
	Data MediaItem `json:"data"`
}

// ListReq lists the uploaded email images.
type ListReq struct {
	g.Meta        `path:"/email-media/list" method:"get" tags:"EmailMedia" summary:"List email images" in:"query"`
	Authorization string `json:"authorization" dc:"Authorization" in:"header"`
}

// ListRes returns the media library, newest first.
type ListRes struct {
	api_v1.StandardRes
	Data struct {
		List []MediaItem `json:"list" dc:"Media items"`
	} `json:"data"`
}

// DeleteReq removes one uploaded image from the library.
type DeleteReq struct {
	g.Meta        `path:"/email-media/delete" method:"post" tags:"EmailMedia" summary:"Delete an email image" in:"body"`
	Authorization string `json:"authorization" dc:"Authorization" in:"header"`
	Name          string `p:"name" v:"required#Please provide the file name" dc:"Stored file name to delete"`
}

// DeleteRes is the response for deleting an image.
type DeleteRes struct {
	api_v1.StandardRes
}
